src/main/java/crawlercommons/sitemaps/SiteMapParserSAX.java

```diff
@@ -211,13 +211,30 @@ public AbstractSiteMap parseSiteMap(String contentType, byte[] content, URL url)
             if (XML_MEDIA_TYPES.contains(mediaType)) {
                 return processXml(url, content);
             } else if (TEXT_MEDIA_TYPES.contains(mediaType)) {
-                return processText(url.toString(), content);
+                return processText(url, content);
             } else if (GZ_MEDIA_TYPES.contains(mediaType)) {
-                return processGzip(url, content);
-            } else {
-                mediaType = MEDIA_TYPE_REGISTRY.getSupertype(mediaType); // Check parent
-                return parseSiteMap(mediaType.toString(), content, url);
+                InputStream decompressed;
+                MediaType embeddedType;
+                try {
+                    decompressed = new GZIPInputStream(new ByteArrayInputStream(content));
+                    embeddedType = MediaType.parse(TIKA.detect(decompressed));
+                } catch (Exception e) {
+                    UnknownFormatException err = new UnknownFormatException("Failed to detect embedded MediaType of gzipped sitemap: " + url + ", caused by " + e);
+                    err.initCause(e);
+                    throw err;
+                }
+                if (XML_MEDIA_TYPES.contains(embeddedType)) {
```

Comment:
I already tried such a solution and decided not to implement it after I've seen a [recursively compressed sitemap file](http://scores.sify.com/sitemap.xml.gz) which would require
- a counter to track and limit recursion to be save from zip bombs
- and also (or alternatively) a parser entry point which unwraps the sitemap recursively

I've found the unrolled solution more clear, esp. because recursion + while-loop looks somewhat transparent. Btw., I've removed the recursive call with the media super type - this is already handled by the while-loop.